File: src/pathfinder_combat_simulator/core/ability_scores.py
from dataclasses import dataclass

@dataclass
class AbilityScores:
    """Represents the six ability scores"""
    strength: int = 10
    dexterity: int = 10
    constitution: int = 10
    intelligence: int = 10
    wisdom: int = 10
    charisma: int = 10

    # Temporary modifiers (from spells, etc.)
    temp_str: int = 0
    temp_dex: int = 0
    temp_con: int = 0
    temp_int: int = 0
    temp_wis: int = 0
    temp_cha: int = 0

    def get_modifier(self, ability: str) -> int:
        """Get the modifier for an ability score"""
        # Map short names to full attribute names
        ability_map = {
            "str": "strength", "dex": "dexterity", "con": "constitution",
            "int": "intelligence", "wis": "wisdom", "cha": "charisma",
            "strength": "strength", "dexterity": "dexterity", "constitution": "constitution",
            "intelligence": "intelligence", "wisdom": "wisdom", "charisma": "charisma"
        }

        temp_map = {
            "str": "temp_str", "dex": "temp_dex", "con": "temp_con",
            "int": "temp_int", "wis": "temp_wis", "cha": "temp_cha",
            "strength": "temp_str", "dexterity": "temp_dex", "constitution": "temp_con",
            "intelligence": "temp_int", "wisdom": "temp_wis", "charisma": "temp_cha"
        }

        full_ability = ability_map.get(ability.lower(), ability)
        temp_attr = temp_map.get(ability.lower(), f"temp_{ability[:3]}")

        total_score = getattr(self, full_ability) + getattr(self, temp_attr)
        return (total_score - 10) // 2

    def get_total_score(self, ability: str) -> int:
        """Get the total ability score including temporary modifiers"""
        # Map short names to full attribute names
        ability_map = {
            "str": "strength", "dex": "dexterity", "con": "constitution",
            "int": "intelligence", "wis": "wisdom", "cha": "charisma",
            "strength": "strength", "dexterity": "dexterity", "constitution": "constitution",
            "intelligence": "intelligence", "wisdom": "wisdom", "charisma": "charisma"
        }

        temp_map = {
            "str": "temp_str", "dex": "temp_dex", "con": "temp_con",
            "int": "temp_int", "wis": "temp_wis", "cha": "temp_cha",
            "strength": "temp_str", "dexterity": "temp_dex", "constitution": "temp_con",
            "intelligence": "temp_int", "wisdom": "temp_wis", "charisma": "temp_cha"
        }

        full_ability = ability_map.get(ability.lower(), ability)
        temp_attr = temp_map.get(ability.lower(), f"temp_{ability[:3]}")

        return getattr(self, full_ability) + getattr(self, temp_attr)

File: src/core/abilityScores.ts
export type Ability = "strength" | "dexterity" | "constitution" | "intelligence" | "wisdom" | "charisma";
export type AbilityShort = "str" | "dex" | "con" | "int" | "wis" | "cha";

export class AbilityScores {
  strength: number;
  dexterity: number;
  constitution: number;
  intelligence: number;
  wisdom: number;
  charisma: number;

  temp_str: number;
  temp_dex: number;
  temp_con: number;
  temp_int: number;
  temp_wis: number;
  temp_cha: number;

  constructor(
    strength: number = 10,
    dexterity: number = 10,
    constitution: number = 10,
    intelligence: number = 10,
    wisdom: number = 10,
    charisma: number = 10,
    temp_str: number = 0,
    temp_dex: number = 0,
    temp_con: number = 0,
    temp_int: number = 0,
    temp_wis: number = 0,
    temp_cha: number = 0,
  ) {
    this.strength = strength;
    this.dexterity = dexterity;
    this.constitution = constitution;
    this.intelligence = intelligence;
    this.wisdom = wisdom;
    this.charisma = charisma;
    this.temp_str = temp_str;
    this.temp_dex = temp_dex;
    this.temp_con = temp_con;
    this.temp_int = temp_int;
    this.temp_wis = temp_wis;
    this.temp_cha = temp_cha;
  }

  private mapAbilityKey(ability: Ability | AbilityShort | string): { baseKey: Ability, tempKey: keyof this } {
    const lowerAbility = ability.toLowerCase();
    switch (lowerAbility) {
      case "str":
      case "strength":
        return { baseKey: "strength", tempKey: "temp_str" };
      case "dex":
      case "dexterity":
        return { baseKey: "dexterity", tempKey: "temp_dex" };
      case "con":
      case "constitution":
        return { baseKey: "constitution", tempKey: "temp_con" };
      case "int":
      case "intelligence":
        return { baseKey: "intelligence", tempKey: "temp_int" };
      case "wis":
      case "wisdom":
        return { baseKey: "wisdom", tempKey: "temp_wis" };
      case "cha":
      case "charisma":
        return { baseKey: "charisma", tempKey: "temp_cha" };
      default:
        throw new Error(`Invalid ability score: ${ability}`);
    }
  }

  getModifier(ability: Ability | AbilityShort | string): number {
    const totalScore = this.getTotalScore(ability);
    return Math.floor((totalScore - 10) / 2);
  }

  getTotalScore(ability: Ability | AbilityShort | string): number {
    const { baseKey, tempKey } = this.mapAbilityKey(ability);
    const baseValue = this[baseKey] as number; // Type assertion
    const tempValue = this[tempKey] as number; // Type assertion
    return baseValue + tempValue;
  }
}
